go/ephemeral/user_ek_box_storage.go

```diff
@@ -0,0 +1,249 @@
+	// is still returned in this case. TODO: Turn this warning into an error
+	// after EK support is sufficiently widespread.
+	if wrongKID {
+		s.G().Log.CWarningf(ctx, "It looks like you revoked a device without generating new ephemeral keys. Are you running an old version?")
+		return userEK, nil
+	}
+	if err != nil {
+		return userEK, err
+	}
+
+	if userEKMetadata == nil { // shouldn't happen
+		s.G().Log.CWarningf(ctx, "No error but got nil userEKMetadata")
+		return userEK, err
+	}
+
+	userEKBoxed := keybase1.UserEkBoxed{
+		Box:                result.Result.Box,
+		DeviceEkGeneration: result.Result.DeviceEKGeneration,
+		Metadata:           *userEKMetadata,
+	}
+
+	userEK, err = s.unbox(ctx, userEKBoxed)
+	if err != nil {
+		return userEK, err
+	}
+
+	seed := UserEKSeed(userEK.Seed)
+	keypair, err := seed.DeriveDHKey()
+	if err != nil {
+		return userEK, err
+
+	}
+
+	if !keypair.GetKID().Equal(userEKMetadata.Kid) {
+		return userEK, fmt.Errorf("Failed to verify server given seed against signed KID %s", userEKMetadata.Kid)
+	}
+
+	// Store the boxed version, return the unboxed
+	err = s.put(generation, userEKBoxed)
+	return userEK, err
+}
+
+func (s *UserEKBoxStorage) Put(ctx context.Context, generation keybase1.EkGeneration, userEKBoxed keybase1.UserEkBoxed) (err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#Put", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+	return s.put(generation, userEKBoxed)
+}
+
+func (s *UserEKBoxStorage) put(generation keybase1.EkGeneration, userEKBoxed keybase1.UserEkBoxed) (err error) {
+	key := s.dbKey()
+	s.cache[generation] = userEKBoxed
+	err = s.G().GetKVStore().PutObj(key, nil, s.cache)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *UserEKBoxStorage) unbox(ctx context.Context, userEKBoxed keybase1.UserEkBoxed) (userEK keybase1.UserEk, err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#unbox", func() error { return err })()
+	deviceEKStorage := s.G().GetDeviceEKStorage()
+	deviceEK, err := deviceEKStorage.Get(ctx, userEKBoxed.DeviceEkGeneration)
+	if err != nil {
+		return userEK, err
+	}
+
+	deviceSeed := DeviceEKSeed(deviceEK.Seed)
+	deviceKeypair, err := deviceSeed.DeriveDHKey()
+	if err != nil {
+		return userEK, err
+	}
+
+	msg, _, err := deviceKeypair.DecryptFromString(userEKBoxed.Box)
+	if err != nil {
+		return userEK, err
+	}
+
+	seed, err := newUserEKSeedFromBytes(msg)
+	if err != nil {
+		return userEK, err
+	}
+
+	return keybase1.UserEk{
+		Seed:     keybase1.Bytes32(seed),
+		Metadata: userEKBoxed.Metadata,
+	}, nil
+}
+
+func (s *UserEKBoxStorage) Delete(ctx context.Context, generation keybase1.EkGeneration) (err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#Delete", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+	delete(s.cache, generation)
+	key := s.dbKey()
+	return s.G().GetKVStore().PutObj(key, nil, s.cache)
+}
+
+func (s *UserEKBoxStorage) GetAll(ctx context.Context) (userEKs UserEKMap, err error) {
```

Comment:
I wonder if we should tweak the name `UserEKMap` to emphasize that it's holding secret key material. It would be really bad if we accidentally saved it to disk somewhere.

Reply:
sure